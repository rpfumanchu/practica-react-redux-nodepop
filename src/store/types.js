export const AUTH_LOGIN_REQUEST = "AUTH_LOGIN_REQUEST";
export const AUTH_LOGIN_SUCCESS = "AUTH_LOGIN_SUCCESS";
export const AUTH_LOGIN_FAILURE = "AUTH_LOGIN_FAILURE";

export const AUTH_LOGOUT = "AUTH_LOGOUT";

export const ADS_LOADED_REQUEST = "ADS_LOADED_REQUEST";
export const ADS_LOADED_SUCCESS = "ADS_LOADED_SUCCESS";
export const ADS_LOADED_FAILURE = "ADS_LOADED_FAILURE";

export const TAGS_LOADED_REQUEST = "TAGS_LOADED_REQUEST";
export const TAGS_LOADED_SUCCESS = "TAGS_LOADED_SUCCESS";
export const TAGS_LOADED_FAILURE = "TAGS_LOADED_FAILURE";

export const AD_CREATED = "AD_CREATED";

export const USERINTERFACE_RESET_ERROR = "USERINTERFACE_RESET_ERROR";

export const TOGGLE_MODAL = "TOGGLE_MODAL";

export const TOGGLE_RESULT = "TOGGLE_RESULT";

export const AD_FILTERING_NAME = "AD_FILTERING_NAME";
export const AD_FILTERING_SALE = "AD_FILTERING_SALE";
export const AD_FILTERING_TAGS = "AD_FILTERING_TAGS";
export const AD_FILTERING_MIN_PRICE = "AD_FILTERING_MIN_PRICE";
export const AD_FILTERING_MAX_PRICE = "AD_FILTERING_MAX_PRICE";
