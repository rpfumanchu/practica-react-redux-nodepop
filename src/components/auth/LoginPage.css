.container-form {
  width: 100%;
  max-width: 500px;
  margin: 120px auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 30px 10px 20px 10px;
  gap: 10px;
  border-radius: 10px;
  text-align: center;
  border: 2px solid var(--main-color);
  background-color: whitesmoke;
}

.container-form .form-label {
  font-size: 20px;
  margin-bottom: 10px;
}

.container-form .form-input {
  height: 30px;
  justify-items: center;
  justify-content: center;
  margin-left: 10px;
  font-size: 23px;
  text-align: center;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}

/* .container-form .form-input + small {
  color: grey;
} */

.button {
  display: block;
  background-color: #578d57;
  border-radius: 5px;
  font-size: 23px;
  color: white;
  border: 0;
  padding: 8px 12px;
  margin-bottom: 10px;
}

.container-form .links {
  display: flex;
  justify-content: center;
  font-size: 23px;
  color: rgba(0, 0, 255, 0.897);
  text-decoration: none;
  margin-bottom: 10px;
}

.loginPage-error {
  background-color: rgb(224, 36, 94, 0.1);
  border: 1px solid rgba(224, 36, 94, 0.2);
  border-radius: 5px;
  color: rgb(224, 36, 94);
  margin-top: 20px;
  padding: 10px 15px;
}
