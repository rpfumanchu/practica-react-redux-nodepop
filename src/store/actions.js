import { AD_LOADED, AUTH_LOGIN, AUTH_LOGOUT } from "./types";

export const authLogin = () => ({
  type: AUTH_LOGIN,
});

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const adLoaded = ad => ({
  type: AD_LOADED,
  payload: ad,
});
