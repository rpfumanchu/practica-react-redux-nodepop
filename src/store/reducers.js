import {
  ADS_LOADED_SUCCESS,
  AD_CREATED_SUCCESS,
  AD_DELETED_SUCCESS,
  AD_FILTERING_MAX_PRICE,
  AD_FILTERING_MIN_PRICE,
  AD_FILTERING_NAME,
  AD_FILTERING_SALE,
  AD_FILTERING_TAGS,
  AD_LOADED_SUCCESS,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  SHOW_MODAL_DELETE,
  TAGS_LOADED_SUCCESS,
  TOGGLE_MODAL,
  TOGGLE_MODAL_DELETE,
  TOGGLE_RESULT,
  USERINTERFACE_RESET_ERROR,
} from "./types";

export const defaultState = {
  auth: false,
  ads: {
    areLoaded: false,
    data: [],
  },
  tags: {
    areLoaded: false,
    data: [],
  },
  userInterface: {
    isLoading: false,
    showModal: false,
    showModalDelete: true,
    showDeleteMessage: null,
    searchResults: true,
    error: null,
  },
  filtered: {
    query: "",
    querySale: "",
    queryTags: [],
    queryMinPrice: "",
    queryMaxPrice: "",
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;

    default:
      return state;
  }
}

export function ads(state = defaultState.ads, action) {
  if (action.type === ADS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  if (action.type === AD_LOADED_SUCCESS) {
    return { ...state, data: [action.payload] };
  }
  if (action.type === AD_DELETED_SUCCESS) {
    const updatedData = state.data.filter(ad => ad.id !== action.payload);
    return { ...state, data: updatedData };
  }
  if (action.type === AD_CREATED_SUCCESS) {
    return { ...state, data: [action.payload].concat(state.data) };
  }

  return state;
}

export function tags(state = defaultState.tags, action) {
  if (action.type === TAGS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  return state;
}

export function filtered(state = defaultState.filtered, action) {
  switch (action.type) {
    case AD_FILTERING_NAME:
      return {
        ...state,
        query: action.payload,
      };
    case AD_FILTERING_SALE:
      return {
        ...state,
        querySale: action.payload,
      };
    case AD_FILTERING_TAGS:
      return {
        ...state,
        queryTags: action.payload,
      };
    case AD_FILTERING_MIN_PRICE:
      return {
        ...state,
        queryMinPrice: action.payload,
      };
    case AD_FILTERING_MAX_PRICE:
      return {
        ...state,
        queryMaxPrice: action.payload,
      };
    default:
      return state;
  }
}

export function userInterface(state = defaultState.userInterface, action) {
  if (action.error) {
    return { isLoading: false, error: action.payload };
  }

  if (/_REQUEST$/.test(action.type)) {
    return { ...state, isLoading: true, error: null };
  }

  if (/_SUCCESS$/.test(action.type)) {
    return { ...state, isLoading: false, showModalDelete: true, error: null };
  }

  if (action.type === USERINTERFACE_RESET_ERROR) {
    return { ...state, error: null };
  }

  if (action.type === TOGGLE_MODAL) {
    return { ...state, showModal: !state.showModal };
  }
  if (action.type === TOGGLE_MODAL_DELETE) {
    return { ...state, showModalDelete: !state.showModalDelete };
  }
  if (action.type === SHOW_MODAL_DELETE) {
    return { ...state, showDeleteMessage: action.value };
  }
  if (action.type === TOGGLE_RESULT) {
    return { ...state, searchResults: action.value };
  }

  return state;
}
