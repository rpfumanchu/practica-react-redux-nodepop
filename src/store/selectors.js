export const getIsLogged = state => state.auth;

export const getAllAds = state => state.ads;

export const getAdId = id => state =>
  getAllAds(state).find(ads => ads.id === id);
