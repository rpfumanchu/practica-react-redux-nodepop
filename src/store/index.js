import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import thunk from "redux-thunk";

import * as auth from "../components/auth/service";
import * as ads from "../components/Ads/service";

import * as actionCreators from "./actions";
import * as reducers from "./reducers";

const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({
  actionCreators,
});

//DONE middleware para manejar las redirecciones cuando hay fallo
const failureRedirects = (router, redirectsMap) => () => next => action => {
  const result = next(action);

  if (action.error) {
    const redirect = redirectsMap[action.payload.status];
    if (redirect) {
      router.navigate(redirect);
    }
  }
  return result;
};

//NOTE a√±ado devtools
//NOTE uso preloadedState para tener un estado precargado
export default function configureStore(preloadedState, { router }) {
  const middleware = [
    thunk.withExtraArgument({ service: { auth, ads }, router }),
    failureRedirects(router, { 401: "/login", 404: "/404" }),
  ];
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware)),
  );
  return store;
}
