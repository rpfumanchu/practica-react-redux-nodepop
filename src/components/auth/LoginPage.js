import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import Button from "../shared/Button";
//import { login } from "./service";
import "./LoginPage.css";
import Layout from "../layout/Layout";
//import { useLocation, useNavigate } from "react-router-dom";
import ErrorModal from "../shared/modal/ErrorModal";
import Spinner from "../shared/spinner/Spinner";
import {
  authLogin,
  toggleModal,
  userInterfaceResetError,
} from "../../store/actions";
import { getUserInterface } from "../../store/selectors";

//DONE Loguear con email y password y un checkbox para dar la opción de persistir el token, además manejar errores y feedback al usuario. Al hacer Login quiero enviar al usuario a la página a la que queria ir.

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, showModal, error } = useSelector(getUserInterface);

  const [credentials, setCredentials] = useState({
    email: "",
    password: "",
    rememberMe: false,
  });

  const resetError = () => {
    dispatch(userInterfaceResetError());
  };

  //const onLogin = () => dispatch(authLoginSuccess());

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    if (type === "checkbox") {
      setCredentials(prevCredentials => ({
        ...prevCredentials,
        [name]: checked,
      }));
    } else {
      setCredentials(prevCredentials => ({
        ...prevCredentials,
        [name]: value,
      }));
    }
  };

  const handleShowModal = () => {
    dispatch(toggleModal());
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));

    dispatch(toggleModal());
  };

  const buttonDisabled =
    isLoading || !credentials.email || !credentials.password;

  return (
    <Layout title="Login Page">
      <div>
        {isLoading ? (
          <Spinner message="cargando..." />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <label className="form-label">Email</label>
            <input
              className="form-input"
              type="text"
              name="email"
              onChange={handleChange}
              value={credentials.email}
              required
            />
            <label className="form-label">Password</label>
            <input
              className="form-input"
              type="password"
              name="password"
              onChange={handleChange}
              value={credentials.password}
              required
            />
            <label>
              Guardar sesiòn
              <input
                type="checkbox"
                name="rememberMe"
                checked={credentials.rememberMe}
                onChange={handleChange}
              />
            </label>
            <Button
              type="submit"
              variant="primary"
              width="button-form"
              disabled={buttonDisabled}>
              Log in
            </Button>
          </form>
        )}

        {showModal && (
          <ErrorModal
            title="Login"
            message={"Acabas de iniciar sesión"}
            onCancel={handleShowModal}
          />
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={resetError}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
