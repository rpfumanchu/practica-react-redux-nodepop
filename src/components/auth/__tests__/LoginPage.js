import { render, screen } from "@testing-library/react";
import LoginPage from "../LoginPage";
import { defaultState } from "../../../store/reducers";
import { authLogin, userInterfaceResetError } from "../../../store/actions";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

jest.mock("../../../store/actions");

describe("LoginPage", () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = defaultState;
        state.userInterface.error = error;

        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>,
    );
  };

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("shoul dispatch authLogin action", () => {
    const email = "rober";
    const password = "123";
    const checkbox = true;

    //NOTE renderizo el componente
    renderComponent();

    const emailInput = screen.getByLabelText(/Email/);

    const passwordInput = screen.getByLabelText(/Password/);

    const checkboxInput = screen.getByLabelText(/rememberMe/);

    const submitButton = screen.getByRole("button");
    //const submitButton = screen.getByTestId("button");

    console.log(submitButton);

    expect(submitButton).toBeDisabled();

    //NOTE para lanzar eventos
    //fireEvent.change(emailInput, { target: { value: email } });

    // fireEvent.change(passwordInput, { target: { value: password } });

    //fireEvent.change(checkboxInput, { target: { checked: checkbox } });

    userEvent.type(emailInput, email);
    userEvent.type(passwordInput, password);
    userEvent.type(checkboxInput, checkbox);

    console.log("averque", email, password, checkbox);

    expect(screen.getByLabelText(/Email/)).toBeInTheDocument();
    expect(screen.getByLabelText(/Password/)).toBeInTheDocument();
    expect(screen.getByLabelText(/rememberMe/)).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();

    //expect(submitButton).toBeEnabled();

    //fireEvent.click(submitButton);

    userEvent.click(submitButton);
    expect(authLogin).toHaveBeenCalledWith({ email, password, checkbox });
  });

  test("should display an error", () => {
    const error = { message: "Unauthorized" };
    renderComponent(error);

    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();

    // fireEvent.click(errorElement);
    userEvent.click(errorElement);

    expect(userInterfaceResetError).toHaveBeenCalled();
  });
});
