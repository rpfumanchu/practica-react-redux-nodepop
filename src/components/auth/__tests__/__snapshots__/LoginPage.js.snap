// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="layout"
  >
    <header>
      <nav
        class="header-navbar layout-header"
      >
        <a
          href="/adverts/home"
        >
          <h4
            class="navbar-h4"
          >
            NodePop
          </h4>
        </a>
        <ul
          class="navbar-list"
        >
          <li
            class="navbar-list-item"
          >
            <a
              class="navbar-list-item"
              href="/adverts/new"
            >
              New
            </a>
            <a
              class="navbar-list-item"
              href="/adverts"
            >
              Latest Ad
            </a>
            <a
              class="sc-beqWaB loJBsH"
              href="/login"
              variant="primary"
            >
              Login
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="title"
      >
        <svg
          class="logo"
        >
          nodepop.svg
        </svg>
        <h1
          class="title-h1"
        >
          Práctica NodePop
        </h1>
        <svg
          class="logo"
        >
          nodepop.svg
        </svg>
      </div>
      <div />
    </header>
    <main
      class="layout-main"
    >
      <h2
        class="layout-title"
      >
        Login Page
      </h2>
      <div>
        <form
          class="container-form"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
          </label>
          <input
            class="form-input"
            id="email"
            name="email"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-input"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <label
            for="rememberMe"
          >
            rememberMe
            <input
              id="rememberMe"
              name="rememberMe"
              type="checkbox"
            />
          </label>
          <button
            class="sc-beqWaB ftSVgc"
            data-testid="button"
            disabled=""
            type="submit"
            width="button-form"
          >
            Log in
          </button>
        </form>
      </div>
    </main>
    <div
      class="footer layout-footer"
    >
      <p
        class="p"
      >
        NodePop®. Todos los derechos reservados 2023
      </p>
      <svg
        class="logo-footer"
      >
        nodepop.svg
      </svg>
    </div>
  </div>
</div>
`;
